buffer/pool.go

```diff
@@ -205,3 +205,58 @@ func (b *Buffer) BuildBytes() []byte {
 
 	return ret
 }
+
+type readCloser struct {
+	offset int
+	bufs   [][]byte
+}
+
+func (r *readCloser) Read(p []byte) (n int, err error) {
+	for _, buf := range r.bufs {
+		// Copy as much as we can.
+		x := copy(p[n:], buf[r.offset:])
+		n += x // Increment how much we filled.
+
+		// Did we empty the whole buffer?
+		if r.offset+x == len(buf) {
+			// On to the next buffer.
+			r.offset = 0
+			r.bufs = r.bufs[1:]
+
+			// We can release this buffer.
+			putBuf(buf)
+		} else {
+			r.offset += x
+		}
+
+		if n == len(p) {
+			break
+		}
+	}
+	// No buffers left or nothing read?
+	if len(r.bufs) == 0 || n == 0 {
```

Comment:
Yes, if you don't return `io.EOF` most readers will never stop waiting for additional data. The `net.Buffers` example you link to returns `io.EOF` as well. Functions like `bytes.Buffer.ReadFrom` [only break on `io.EOF`](https://golang.org/src/bytes/buffer.go?s=5754:5813#L182).